plugins/l4lb/l4lb.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2015 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+
+	"github.com/dcos/dcos-cni/pkg/spartan"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/invoke"
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+)
+
+type NetConf struct {
+	types.NetConf
+	Args     map[string]interface{} `json:"args"`
+	MTU      int                    `json:"mtu"`
+	Delegate map[string]interface{} `json:"delegate"`
+}
+
+func init() {
+	// This ensures that main runs only on main thread (thread group leader).
+	// Since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func setupDelegateConf(conf *NetConf) (delegateConf []byte, delegatePlugin string, err error) {
+	conf.Delegate["name"] = conf.Name
+	conf.Delegate["cniVersion"] = conf.CNIVersion
+	conf.Delegate["args"] = conf.Args
+
+	delegateConf, err = json.Marshal(conf.Delegate)
+	if err != nil {
+		err = fmt.Errorf("failed to marshall the delegate configuration: %s", err)
+		return
+	}
+
+	delegatePlugin, ok := conf.Delegate["type"].(string)
```

Comment:
conf.Delegate["type"] tries to retrieve the value associated with the key "type". That has 2 return args, the value and a boolean that indicated whether or not the key was found.

value.(string) is doing a type assertion. That has 2 return args, the asserted value and a boolean that indicated whether or not the conversion was valid.

These are 2 unrelated failure conditions and should be separated out. Unless we can guarantee that conf.Delegate["type"] will always succeed (I don't think we can).